ai_ta_backend/cost_tracking.py

```diff
@@ -0,0 +1,225 @@
+        return "Success!"
+    except Exception as e:
+        print("Error in cost calculation and logging: ", e)
+        return "Failed!"
+
+def get_tokens_and_cost(text: str, std_prompt: str, model_name: str, flag: str = 'query'):
+    """
+    Takes an input query and returns the cost.
+    1. Cost of encoding query + prompt
+    2. Cost of answer generation by the model
+    """
+    #print("in get_tokens_and_cost")
+    # initialize tokenizer
+    tokenizer = tiktoken.encoding_for_model("gpt-3.5-turbo")
+    if model_name.startswith('gpt-4'):
+        if "32k" in model_name:
+            # 32k context
+            prompt_cost_per_token: float = 0.06/1000
+            completion_cost_per_token: float = 0.12/1000
+        else:
+            # 8k context
+            prompt_cost_per_token: float = 0.03/1000
+            completion_cost_per_token: float = 0.06/1000
+
+    elif model_name.startswith('gpt-3.5-turbo'):
+        if "16k" in model_name:
+            # 16k context
+            prompt_cost_per_token: float = 0.003/1000
+            completion_cost_per_token: float = 0.004/1000
+        else:
+            # 4k context
+            prompt_cost_per_token: float = 0.0015/1000
+            completion_cost_per_token: float = 0.002/1000
+
+    else:
+        print(f"No idea of cost! Pricing not supported for model: `{model_name}`")
+
+    if flag == 'query':
+        input_str = std_prompt + " " + text
+        num_of_input_tokens = len(tokenizer.encode(input_str))
+        prompt_cost = round(float(prompt_cost_per_token * num_of_input_tokens),4)
+        return prompt_cost, num_of_input_tokens
+    else:
+        output_str = text
+        num_of_output_tokens = len(tokenizer.encode(output_str))
+        response_cost = round(float(completion_cost_per_token * num_of_output_tokens),4)
+        return response_cost, num_of_output_tokens
+
+
+def get_course_cost(course_name: str, days=0):
+    """
+    Returns a dataframe of conversations for the last given days and course.
+    If days=0, return all conversations for the course.
+    """
+    # initialize supabase
+    supabase_client = supabase.create_client(supabase_url=os.getenv('SUPABASE_URL'), 
+                                             supabase_key=os.getenv('SUPABASE_API_KEY'))
+    try:
+        if days > 0:
+            # fetch conversations for the last given days
+            current_date = datetime.now()
+            days_ago = current_date - timedelta(days=days)
+            given_date = days_ago.strftime("%Y-%m-%d")
+            print("Current Date:", current_date.strftime("%Y-%m-%d"))
+            print("30 Days Ago:", given_date)
+            response = supabase_client.table("llm-convo-monitor").select("id", count='exact').eq("course_name", course_name).gte('created_at', given_date).order('id', desc=False).execute()
```

Comment:
👍 Great workaround to limit the size of the supabase call... Hopefully we never have a single day with tons of conversations that will overflow the DB memory.